src/main/java/software/amazon/awssdk/s3accessgrants/plugin/S3AccessGrantsIdentityProvider.java

```diff
@@ -124,27 +121,29 @@ public CompletableFuture<? extends AwsCredentialsIdentity> resolveIdentity(Resol
         CompletableFuture<? extends AwsCredentialsIdentity> userCredentials = null;
 
         try {
-            String accountId = getCallerAccountID().join().account();
-
-            validateRequestParameters(resolveIdentityRequest, accountId, privilege, isCacheEnabled);
 
             userCredentials = credentialsProvider.resolveIdentity(resolveIdentityRequest);
 
+            if(resolveIdentityRequest != null && resolveIdentityRequest.property(AUTH_EXCEPTIONS_PROPERTY) != null) {
+                throw (SdkServiceException) resolveIdentityRequest.property(AUTH_EXCEPTIONS_PROPERTY);
+            }
+
+            String accountId = getCallerAccountID().join().account();
+            validateRequestParameters(resolveIdentityRequest, accountId, privilege, isCacheEnabled);
+
             String S3Prefix = resolveIdentityRequest.property(PREFIX_PROPERTY).toString();
-            String operation = resolveIdentityRequest.property(OPERATION_PROPERTY).toString();
+            Permission permission = Permission.fromValue(resolveIdentityRequest.property(PERMISSION_PROPERTY).toString());
+            Region destinationRegion = Region.of(resolveIdentityRequest.property(BUCKET_LOCATION_PROPERTY).toString());
+
+            S3ControlAsyncClient s3ControlAsyncClient = s3ControlBuilder.region(destinationRegion).build();
```

Comment:
Are we creating new s3ControlClients every time a new request is made? Can we have a map where we store the S3ControlClient we already created and reuse them for the same region.

Reply:
Yes, we are creating a builder that will be passed onto the identity providers during plugin creation. This builder will have all the properties set except for region. Region will be set once the bucket region is available and a client will be built.

Reply:
Even if the Builder object is reused, we are creating a new S3ControlAsyncClient object for each request, even if it it to the same region. I think we should store existing clients and reuse them when possible.

Reply:
We can certainly use maps to store the clients but I did not see a benefit with caching the clients. The reason being that we are doing the heavy work of initializing all the parameters except region during plugin initialization. building a client with all the parameters already initialized should be quick enough (as the build process is not complicated). So, the idea was that we can initialize the client for each request but if we deem that the initialization is adding additional overhead, we can switch to using the maps (in future versions). However, if you can think of scenarios where using the map will be more beneficial I can work on adding caching for the clients.